// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package private_statistics.prio.proto;

import "prio/proto/algorithm_parameters.proto";

option java_multiple_files = true;

// Represents a batch of validity share packets for the Prio algorithm, as
// periodically produced by the aggregation servers.
message PrioValidityShareBatch {
  // Universal unique identifier of the PrioDataShareBatch to which batch
  // corresponds to.
  optional string batch_uuid = 1;
  // Parameters for this Prio aggregation.
  optional PrioAlgorithmParameters parameters = 2;
  // The collection of individual validity share packets
  repeated PrioValiditySharePacket packets = 3;
  // A name for this specific aggregation.
  optional string name = 4;
  // Signature of data.
  optional bytes signature_of_packets = 5;
  // Some identifier as to which key was used to sign
  optional string signature_id = 6;
}

// An individual validity share packet created by a Prio aggregation server.
message PrioValiditySharePacket {
  // Universal unique identifier of the PrioDataSharePacket to which this
  // validity share corresponds to.
  optional string uuid = 1;
  // The value of the validity share.
  optional PrioValidityShare validity_share = 2;
}

// The specific polynomial coefficients of the validity share.
message PrioValidityShare {
  optional uint64 f_r = 1;
  optional uint64 g_r = 2;
  optional uint64 h_r = 3;
}
